### PR TITLE
Add GCS signed URL support

### LINKED ISSUE #5233: GCS Signed URL Support
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

https://github.com/apache/arrow-rs/pull/4876 added support for generating S3 URLs, via the [Signer](https://docs.rs/object_store/latest/object_store/signer/trait.Signer.html) trait. This ticket tracks implementing Signer for [GoogleCloudStorage](https://docs.rs/object_store/latest/object_store/gcp/struct.GoogleCloudStorage.html)

**Describe the solution you'd like**

The process for generating signed URLs is described [here](https://cloud.google.com/storage/docs/access-control/signed-urls).

Once the stringToSign has been constructed there are [two mechanisms](https://cloud.google.com/storage/docs/access-control/signing-urls-manually) for generating the signature:

* Directly sign the URL using the RSA key pair of a service account
* Make an authorized API call to the [signBlob](https://cloud.google.com/storage/docs/authentication/creating-signatures) API

The latter approach will support all [GcpCredentialProvider](https://docs.rs/object_store/latest/object_store/gcp/type.GcpCredentialProvider.html) and is therefore probably the approach to start with. The former is a touch more fiddly, and will likely involve some rejigging of GoogleCloudStorageBuilder::build to expose this ServiceAccountKey in such a way that it can be used by the Signer implementation. It would be perfectly acceptable for the first version to only support the signBlob approach.

**Describe alternatives you've considered**

**Additional context**

Split out of https://github.com/apache/arrow-rs/issues/3027

### PR DESCRIPTION
**Pull request by l1nxy**
# Which issue does this PR close?

Closes #5233.

# Rationale for this change
 

Adding support for GCS Signed URL generation.
# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by l1nxy** [comment]
@tustvold Hi, I have completed the first version of GCP signed URL generation. Could you please take some time to review it?  and then i can update the code gradually.
Regarding the credential structure, I think we need to add the client email field. This is because the signed URL generation requires the client email, not just the sign blob method. Even when generating a signed URL with an HMAC key, we still need the client email to create credentials with scope,  the code example showed.

https://github.com/GoogleCloudPlatform/python-docs-samples/blob/6e591bec774bf0f3bf53864146f24028b98ac8df/storage/signed_urls/generate_signed_urls.py#L63-L65

next, i will add test for the sign blob and add code for using HAMC key to generating signed URL.

**Comment by tustvold** [comment]
Sorry for the delay in reviewing this, I wonder if instead of modifying `GcpCredential` which is a breaking change, we could introduce a `GcpSigningCredential` with something like

```
struct GcpSigningCredential {
    credential: GcpCredential,
    email: String,
}
```

This could then have a separate `CredentialProvider<GcpSigningCredential>`

**Comment by l1nxy** [comment]
I completely agree with you. However, a question arises: if I add a new credential type, will all implementations for TokenProvider need to be rewritten for GcpSigningCredential? This is because it involves additional operations to retrieve the client's email address. I am wondering if this would be acceptable.

**Comment by tustvold** [comment]
I would think you could use composition to avoid code duplication, with the GcpSigningCredential implementation calling through to the GcpCredential implementation

**Comment by s-i-l-k-e** [comment]
@l1nxy how is this going? Let me know if there is anything I can do to help -- would love to use this!

**Comment by l1nxy** [comment]
> @l1nxy how is this going? Let me know if there is anything I can do to help -- would love to use this!

Apologies for the delayed update. I have been busy and unable to dedicate time to this. If you would like to proceed, feel free to do so, and I will close the pr.

**Comment by l1nxy** [comment]
@s-i-l-k-e Hi, I have time now and will continue working on this.

**Comment by l1nxy** [comment]
@tustvold   Hi, sorry for the delayed update, i have a question and hope you can help me. my question is which part modification will cause a breaking change, can i add GcpSigningCredential provider to GoogleCloudStorageBuilder and GoogleStorageConfig?

**Comment by tustvold** [comment]
So long as you don't change an existing public API it won't be a breaking change. So you could add additional fields or new methods, but not change existing ones

**Comment by l1nxy** [comment]
@tustvold Hi, I tested it using my GCP account. The ServiceAccount approach yielded correct results, but testing the email from the meta server was not done. Perhaps adding more unit tests or addressing some code issues. maybe i can add sign by HMAC key.

the test result is: 
Snipaste_2024-04-01_18-46-52

**Review by tustvold — COMMENTED** [review]
This is looking really nice, I took the liberty of pushing some relatively minor API tweaks. I'll test this out tomorrow, and assuming that all works get it merged

File: object_store/src/gcp/credential.rs [hunk 1]
```diff
@@ -689,10 +692,12 @@ impl GCSAuthorizer {
             .append_pair("X-Goog-SignedHeaders", &signed_headers);
 
         let string_to_sign = self.string_to_sign(date, &method, url, &headers);
-        let signature = match &self.credential.private_key {
-            Some(key) => key.sign(&string_to_sign)?,
-            None => client.sign_blob(&string_to_sign, email).await?,
-        };
+        // let signature = match &self.credential.private_key {
```
Comment (tustvold) [inline object_store/src/gcp/credential.rs]: ?

File: object_store/src/gcp/credential.rs [hunk 2]
```diff
@@ -44,7 +44,10 @@ use std::time::{Duration, Instant};
 use tracing::info;
 use url::Url;
 
-pub const DEFAULT_SCOPE: &str = "https://www.googleapis.com/auth/devstorage.full_control";
+pub const DEFAULT_SCOPE: [&str; 2] = [
+    "https://www.googleapis.com/auth/devstorage.full_control",
+    "https://www.googleapis.com/auth/cloud-platform",
```
Comment (tustvold) [inline object_store/src/gcp/credential.rs:49]: What is the motivation for this change?

**Comment by tustvold** [comment]
I've tested this and it appears to be working correctly :tada: 

Thank you for this, epic work

**Comment by alamb** [comment]
Thank you @l1nxy  and @tustvold  -- this is a very nice feature
